python-sdk/nuscenes/eval/panoptic/merge_lidarseg_and_tracking.py

```diff
@@ -0,0 +1,167 @@
+    :param eval_set: Which dataset split to evaluate on, train, val or test.
+    :param out_dir: Path to save any output to.
+    :param verbose: Whether to print any output.
+    """
+    sample_tokens = get_samples_in_eval_set(nusc, eval_set)
+    num_samples = len(sample_tokens)
+    mapper = PanopticClassMapper(nusc)
+    coarse2idx = mapper.get_coarse2idx()
+    if verbose:
+        print(f'There are {num_samples} samples.')
+
+    panoptic_subdir = os.path.join('panoptic', eval_set)
+    panoptic_dir = os.path.join(out_dir, panoptic_subdir)
+    os.makedirs(panoptic_dir, exist_ok=True)
+
+    # Load the predictions.
+    pred_boxes_all, meta = load_prediction(track_json,
+                                           1000,
+                                           TrackingBox,
+                                           verbose=verbose)
+    pred_boxes_all = add_center_dist(nusc, pred_boxes_all)
+
+    inst_tok2id = {}
+    for sample_token in tqdm(sample_tokens, disable=not verbose):
+        sample = nusc.get('sample', sample_token)
+        scene_token = sample['scene_token']
+        if scene_token not in inst_tok2id:
+            inst_tok2id[scene_token] = {}
+
+        # Get the sample data token of the point cloud.
+        sd_record = nusc.get('sample_data', sample['data']['LIDAR_TOP'])
+        cs_record = nusc.get('calibrated_sensor', sd_record['calibrated_sensor_token'])
+        pose_record = nusc.get('ego_pose', sd_record['ego_pose_token'])
+
+        lidar_path = os.path.join(nusc.dataroot, sd_record['filename'])
+        # Load the predictions for the point cloud.
+        lidarseg_pred_filename = os.path.join(seg_folder, 'lidarseg',
+                                              nusc.version.split('-')[-1], sd_record['token'] + '_lidarseg.bin')
+        lidar_seg = LidarSegPointCloud(lidar_path, lidarseg_pred_filename)
+
+        panop_labels = np.zeros(lidar_seg.labels.shape, dtype=np.uint16)
+        overlaps = np.zeros(lidar_seg.labels.shape, dtype=np.uint8)
+
+        pred_boxes = pred_boxes_all[sample_token]
+
+        sorted_pred_boxes, pred_cls, tracking_ids = sort_confidence(pred_boxes)
+        sorted_pred_boxes = boxes_to_sensor(sorted_pred_boxes, pose_record, cs_record)
+
+        for instance_id, (pred_box, cl, tracking_id) in enumerate(zip(sorted_pred_boxes, pred_cls, tracking_ids)):
+            cl_id = coarse2idx[cl]
+            if not inst_tok2id[scene_token]:
+                inst_tok2id[scene_token][tracking_id] = 1
+            elif tracking_id not in inst_tok2id[scene_token]:
+                inst_tok2id[scene_token][tracking_id] = len(inst_tok2id[scene_token]) + 1
+            msk = np.zeros(lidar_seg.labels.shape, dtype=np.uint8)
+            indices = np.where(points_in_box(pred_box, lidar_seg.points[:, :3].T))[0]
+
+            msk[indices] = 1
+            msk[np.logical_and(lidar_seg.labels != cl_id, msk == 1)] = 0
+            intersection = np.logical_and(overlaps, msk)
```

Comment:
after the `logic_and`, would it be possible that `np.sum(msk) == 0`, if so,  need to check it to avoid zero-division.

Reply:
Good point, I've added a small non-value (`1e-32`) to prevent zero-divison